transports/quic/src/connection.rs

```diff
@@ -0,0 +1,389 @@
+    ///
+    /// On success, a [`quinn_proto::StreamEvent::Finished`] event will later be produced when the
+    /// substream has been effectively closed. A [`ConnectionEvent::StreamStopped`] event can also
+    /// be emitted.
+    pub fn finish_substream(
+        &mut self,
+        id: quinn_proto::StreamId,
+    ) -> Result<(), quinn_proto::FinishError> {
+        self.connection.send_stream(id).finish()
+    }
+
+    /// Polls the connection for an event that happened on it.
+    pub fn poll_event(&mut self, cx: &mut Context<'_>) -> Poll<ConnectionEvent> {
+        let mut closed = None;
+        loop {
+            match self.from_endpoint.poll_next_unpin(cx) {
+                Poll::Ready(Some(event)) => {
+                    self.connection.handle_event(event);
+                    continue;
+                }
+                Poll::Ready(None) => {
+                    if closed.is_none() {
+                        return Poll::Ready(ConnectionEvent::ConnectionLost(
+                            ConnectionError::EndpointDriverCrashed,
+                        ));
+                    }
+                }
+                Poll::Pending => {}
+            }
+
+            // Sending the pending event to the endpoint. If the endpoint is too busy, we just
+            // stop the processing here.
+            // We need to be careful to avoid a potential deadlock if both `from_endpoint` and
+            // `to_endpoint` are full. As such, we continue to transfer data from `from_endpoint`
+            // to the `quinn_proto::Connection` (see above).
+            // However we don't deliver substream-related events to the user as long as
+            // `to_endpoint` is full. This should propagate the back-pressure of `to_endpoint`
+            // being full to the user.
+            if let Some(to_endpoint) = self.pending_to_endpoint.take() {
+                match self.endpoint_channel.try_send(to_endpoint, cx) {
+                    Ok(Ok(())) => {}
+                    Ok(Err(to_endpoint)) => {
+                        self.pending_to_endpoint = Some(to_endpoint);
+                        return Poll::Pending;
+                    }
+                    Err(_) => {
+                        return Poll::Ready(ConnectionEvent::ConnectionLost(
+                            ConnectionError::EndpointDriverCrashed,
+                        ));
+                    }
+                }
+            }
+
+            // The maximum amount of segments which can be transmitted in a single Transmit
+            // if a platform supports Generic Send Offload (GSO).
+            // Set to 1 for now since not all platforms support GSO.
+            // TODO: Fix for platforms that support GSO.
+            let max_datagrams = 1;
+            // Poll the connection for packets to send on the UDP socket and try to send them on
+            // `to_endpoint`.
+            if let Some(transmit) = self.connection.poll_transmit(Instant::now(), max_datagrams) {
```

Comment:
I read the docs as in "at least in these cases it must be polled", not that it is **only** allowed in these cases. But I agree that we should not unnecessary poll. However, I think right now we only poll additionally:
-  Once per `Connection::poll_event` call: I think this is okay given that `Connection` is polled in its own task only when something made progress (some I/O happened).
- After `poll_transmit` returned `Some` we poll it again -> I think that make sense because there could be multiple transmits to send out
- After we sent an `EndpointEvent` to the endpoint. Looking into the code it seems like this only happens once in a while if the connection id, or the reset token changes. Also, the endpoint may return an event back where we then call `handle_event`.

Are there any other cases that I am missing / do you think any of these result in too much polling?